src/blog/first-100k-lines.vmd

```diff
@@ -0,0 +1,415 @@
+
+
+Languages like Vale and Rust are particularly strong in this regard because of their [type-state programming](http://cliffle.com/blog/rust-typestate/) [# Type-state programming is a way for the compiler to ensure that a particular class doesn't fall into an invalid state.]
+and especially Vale's [Higher RAII](https://verdagon.dev/blog/higher-raii-7drl). [# Higher RAII will make sure you never forget to call a particular function.]
+
+
+# A better way to comment
+
+In 2019, I came across some of our code for generational references. [# [Generational references](https://vale.dev/blog/generational-references) are an alternative mechanism for memory safety, like reference counting, garbage collection, or borrow checking.]
+
+"That's odd," I said, "This really should be simpler." and I started refactoring.
+
+However, I quickly ran into a wall, and discovered why our code looked odd: it was actually handling some pretty complicated requirements that required that odd approach.
+
+
+In 2021, I came across the same code.
+
+"That's odd," I said, "This really should be simpler." and I started refactoring.
+
+...and proceeded to run into _that same wall._ It was only then that I remembered my first attempt.
+
+
+*This is why we leave comments!*
+
+
+However, comments can become out-of-date, and you might not be lucky enough to stumble across the right comment before you embark on a refactoring adventure.
+
+
+For this reason, we also scatter *clues* around, like the below `see SAIRFU`.
+
+```
+// Here we used to check that the rule's runes were solved, but
+// we don't do that anymore because some rules leave their runes
+// as mysteries, see SAIRFU.
+```
+
+
+
+"See (acronym)" means to search our internal docs for more explanations and discussion. In this case, the documentation's SAIRFU section says:
+
+```
+# Send and Impl Rules For Upcasts (SAIRFU)
+
+`a MyInterface = MyStruct(1337);`
+
+Let's say a's coord rune is X. We can't say:
+`X = MyInterface`
+`X = MyStruct`
+because that's a conflict.
+
+We could just not feed the X = MyStruct into the solver, but then this breaks:
+`a = MyStruct(1337);`
+because it can't figure out what X should be.
+
+...
+```
+
+
+
+This is good for many reasons:
+
+ * We can paste "Some rules leave their runes as mysteries, see SAIRFU." into many comments across our codebase, anywhere that's relevant. This is much better than copying an entire lengthy explanation to a dozen places in the code.
+ * If we later decide SAIRFU wasn't wise, we can search the codebase for SAIRFU to find all of the places we need to update.
+
+
+*TL;DR:* Have many links to centralized documentation.
+
+
+
+# Avoid Nondeterminism
+
+Have you ever had a bug that you could only reproduce one out of every ten tries? What an adventure!
+
+And then there's the bugs that you might only reproduce every hundredth try... *a harrowing endeavor.* You try and you try, and eventually give up and close the bug report as "Cannot Reproduce".
```

Comment:
Honestly ellipsis are abused to much, looking into rephrasing things to avoid them.

Reply:
Wise, I definitely overuse them.